extensions/binding/org.eclipse.smarthome.binding.tradfri/src/main/java/org/eclipse/smarthome/binding/tradfri/handler/TradfriGatewayHandler.java

```diff
@@ -77,20 +85,53 @@ public void handleCommand(ChannelUID channelUID, Command command) {
     @Override
     public void initialize() {
         TradfriGatewayConfig configuration = getConfigAs(TradfriGatewayConfig.class);
-        if (configuration.host == null || configuration.host.isEmpty()) {
+
+        if (isNullOrEmpty(configuration.host)) {
             updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
                     "Host must be specified in the configuration!");
             return;
         }
-        if (configuration.code == null || configuration.code.isEmpty()) {
-            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
-                    "Security code must be provided in the configuration!");
-            return;
+        if (isNullOrEmpty(configuration.code)) {
+            if (isNullOrEmpty(configuration.identity) || isNullOrEmpty(configuration.preSharedKey)) {
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
+                        "Either security code or identity and pre-shared key must be provided in the configuration!");
+                return;
+            }
+        } else {
+            if (isOldFirmware()) {
+                /*
+                 * older firmware - fall back to authentication with security code
+                 * in this case the Thing configuration will not be persisted
+                 */
+                logger.warn("Gateway with old firmware - please consider upgrading to the latest version.");
+                configuration.identity = "";
+                configuration.preSharedKey = configuration.code;
+            } else {
+                configuration.identity = UUID.randomUUID().toString().replace("-", "");
+                configuration.preSharedKey = authenticateOnGateway(configuration);
+
+                if (isNullOrEmpty(configuration.preSharedKey)) {
+                    updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, "Pre-shared key was not obtain successfully");
```

Comment:
What happened to my comments that there are many different reasons you have to treat differently? (authentication failed because of an invalid code, authentication failed because gateway is not reachable, authentication failed because device at the other end does not seem to be a Tradfri gateway, authentication failed due to some internal issue on the Tradfri gateway?). You will need to set different Thing statuses for these cases.

Reply:
> (authentication failed because of an invalid code, authentication failed because gateway is not reachable, authentication failed because device at the other end does not seem to be a Tradfri gateway, authentication failed due to some internal issue on the Tradfri gateway?)

I don't think all does statuses can be distinguished. It might work for ```authentication failed because gateway is not reachable``` but for the others ... not so sure.

And I guess you won't like throwing an exception in the ```authenticateOnGateway``` and catching it in the initialize :)

and again - looking at the current version - this is not done anyway.

Reply:
> And I guess you won't like throwing an exception in the authenticateOnGateway and catching it in the initialize :)

This is exactly what I would expect. The "lucky flow" is that the method returns a psk. For error cases, we must provide some information and `null` isn't a good one. Hence exceptions should be the way to go for proper error handling.

Reply:
OK - will update it to use exceptions